### DISCUSSION TITLE
Precision of unit change (temperature degC to degF)

### DISCUSSION THREAD

**Discussion by maverick56** [josdejong/mathjs#2888] [Q&A]
Hello,

I'm surprised about different precision when using different method to change unit in mathjs.

1/ the expression to change unit degree celcius to fahrenheit is this one :
(VAL °C × 9/5) + 32 = RESULT °F

so it's possible to obtain exact value.

2/ I configure mathjs like this : 
`
const math: math.MathJsStatic = create(all, {
        epsilon: 1e-20,
        number: "Fraction",
      });
`

3/ the goal : transform 60°C and get 140°F

1rst test : 
`math.unit(60, "degC").toNumber("degF")`
I get 139.999...

2nd test : 
`math.unit(60, "degC").toNumeric("degF")`
I get 139.999...

3rd test :
`math.evaluate("60 degree to fahrenheit")`
I get 140 !

does someone explain that ?
why only evaluate give the exact result ?

for information evaluate took 0.030ms, toNumber or toNumeric 0.007ms ...

thanks for answer.

**Comment by josdejong** [comment]
In your examples, your input is a `number` with value `60`. In that case, mathjs will output a `number` for you. I think what you want is:

```js
math.unit(math.fraction(60), "degC").toNumber("degF")
```

**Comment by maverick56** [comment]
tks for your very quick answer !

In fact i was expected the math configuration was enought ...
I doen't want to change all functions i waw writen if i change the configuration.

I try to test your example, it's work. I obtain the exact result. 
But i have typescript warnings ...

could you explain the performance difference with evalute ?
in this case it's better with 0.014ms ...

**Reply by josdejong** [reply]
If you run `math.evaluate(...)`, it will first parse the string and then evaluate it. That is more work then just evaluating it. 

If you want to optimize evaluation, you can parse once, and evaluate the parsed expression repeatedly with a different set of variables in the scope, you can look up the documentation to figure out how that works.

**Comment by maverick56** [comment]
tks again for answer.

I cannot optimized because unit will change each time i will use it. And i cannot add unit in the evaluate scopes ...

You sould add math.fraction type in unit parameter (only number and bigNumber authorized today).
The best will by if there is fraction in math configuration, no depends the unit parameter, it sould wokd with fractions ...

Thanks

**Reply by josdejong** [reply]
Ah, now I see the TypeScript definitions are out of date. Fixed that via 391677f9e0c7cdb39e7511eb48e082a8174cf708 (not yet published)

**Reply by josdejong** [reply]
Fixed definitions are now published in `v11.6.0`.
